KAFKA-15174: Ensure CommitAsync propagate the exception to the user

philipnee opened the pull request:
The PR covers a few important points:
1. Exception handling: We should be thrown RetriableCommitException when the commit exception is retriable.  We should throw FencedIdException on commit and poll similar to the current implementation.  Other errors should be thrown as it is.
2. Callback invocation: The callbacks need to be invoked on the main/application thread; however, the future is completed in the background thread.  To achieve this, I created an Invoker class with a queue, so that these callback can be invoked during the consumer.poll()

Note: One change I made is to remove the DefaultOffsetCommit callback.  Since the callback is purely for logging, I think it is reasonable for us to move the logging to the background thread instead of relying on the application thread to trigger the logging.

philipnee committed:
ensure commitAsync throws the right exception

kirktrue commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java:
```diff
@@ -384,12 +384,10 @@ public void commitAsync(Map<TopicPartition, OffsetAndMetadata> offsets, OffsetCo
         final OffsetCommitCallback commitCallback = callback == null ? new DefaultOffsetCommitCallback() : callback;
         future.whenComplete((r, t) -> {
             if (t != null) {
-                commitCallback.onComplete(offsets, new KafkaException(t));
+                commitCallback.onComplete(offsets, (Exception) t);
             } else {
                 commitCallback.onComplete(offsets, null);
             }
-        }).exceptionally(e -> {
-            throw new KafkaException(e);
```
I had a question about this code (even from before this change)—On which thread will the user's callback be executed?

I'm pretty sure that the event's `Future` is completed on the background thread, right? If so, that means that the callback passed to `whenComplete` is going to be completed on the background thread. And since it then calls the user's callback, it will also be called on the background thread.

Thoughts?

philipnee replied:
thanks for catching this. I think you are right.

philipnee committed:
Ensure fencedInstance is being thrown

philipnee committed:
callback invoker

Update PrototypeAsyncConsumerTest.java

Ensure callback is executed on the application thread

Revert "Update PrototypeAsyncConsumerTest.java"

This reverts commit a76f801503955547d9a0a5ae602b6ed29227f0bd.

implemented callback invocation

clean up

clean up

clean up

dajac commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java:
```diff
@@ -154,6 +158,8 @@ public class PrototypeAsyncConsumer<K, V> implements Consumer<K, V> {
     private AtomicBoolean fencedInstance = new AtomicBoolean(false);
     private final Optional<String> groupInstanceId;
 
+    private ExecutorService callbackExecutor = Executors.newSingleThreadExecutor();
```
This is a bit weird, no? In the current model, I suppose that the callbacks are executed by the application thread when poll() is called. Should we do the same here?

philipnee replied:
Sorry - this is wrong.  The way I wanted to do it in this PR is to register the callbacks, when the future is completed, to the application thread.  And we invoke them when we call poll.  I believe the callback invocation is currently invoked on consumer.poll(), because this is where network clients are polled.

dajac commented:
Does commitSync have the same issue?

dajac commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java:
```diff
@@ -1096,4 +1114,47 @@ private void subscribeInternal(Collection<String> topics, Optional<ConsumerRebal
+     */
+    private static class CallbackInvoker {
+        // Thread-safe queue to store callbacks
+        private final BlockingQueue<Runnable> callbackQueue = new LinkedBlockingQueue<>();
```
I also wonder if we really need this new queue. Could we rely on the existing queue between the background and the foreground threads? I haven't look into it myself but I wonder...

philipnee replied:
Possible.  However, the OffsetCommitCallback was not registered to the background thread, that why I added an invoker.

philipnee replied:
In fact - using the backgroundEventQueue would not be feasible. I just realized the callbacks need to be invoked on all commit calls, and poll.  I don't know if there's a better way to do it.

kirktrue commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java:
```diff
@@ -381,20 +394,33 @@ public void commitAsync(OffsetCommitCallback callback) {
+                }
+            };
+            invoker.submit(task);  // Store the callback task for future execution
         });
```
The code passed to `whenComplete` will be run on the background thread. Therefore, it's "acceptable" (IMO) to enqueue a new event onto the background event queue. Then, when it's later run in the background event processor, the event processing should trigger the code you have here.

kirktrue commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java:
```diff
@@ -326,6 +337,8 @@ public PrototypeAsyncConsumer(LogContext logContext,
     @Override
     public ConsumerRecords<K, V> poll(final Duration timeout) {
+        maybeInvokeCallbacks();
+        maybeThrowFencedInstanceException();
```
Ideally, these checks should be added to the logic in `BackgroundEventProcessor`. Right now, `BackgroundEventProcessor` is a separate class, which makes it a bit awkward, since you need access to the consumer's private instance variables. In a PR that I have in review, the `BackgroundEventProcessor` is an inner class of the consumer, making getting and setting the instance variable values trivial.

kirktrue commented on clients/src/test/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumerTest.java:
```diff
@@ -168,6 +200,45 @@ public void testCommitted_ExceptionThrown() {
+
+        @Override
+        public void onComplete(Map<TopicPartition, OffsetAndMetadata> offsets, Exception exception) {
+            System.out.println("invoke");
```
We should remove the `System.out.println()` from committed code.

philipnee replied:
Java debugging 101: use println

philipnee committed:
clean up the callback tasks

philipnee committed:
Test exception handling

philipnee commented:
Hi @dajac @kirktrue - Thanks for offering valuable feedbacks to this PR, much appreciated. I wonder how do you feel about using a local queue to stash the new callbacks? The goal really is to ensure these callbacks can be invoked on commit, poll, and close calls.  Previously, David suggested that we could try to use the background queue; however, this is not really a feasible solution because we don't want the consumer to poll the queue during a commit (which could spit out a bunch of errors and rebalance events).

I don't have a clean way of doing this unless we make a KIP change to enforce callbacks only invoked on poll.  This is idea, but probably out of the scope.  Any suggestions?

philipnee committed:
ensure test pass

updating the IT to pass

philipnee committed:
Update documentation

lucasbru commented on core/src/test/scala/integration/kafka/api/BaseAsyncConsumerTest.scala:
```diff
@@ -16,7 +16,7 @@
 package kafka.api
 
-import kafka.utils.TestUtils.waitUntilTrue
+import kafka.utils.TestUtils.{waitUntilTrue}
```
Do we need curly brackets?

kirktrue commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java:
```diff
@@ -147,6 +152,10 @@ public class PrototypeAsyncConsumer<K, V> implements Consumer<K, V> {
     // to keep from repeatedly scanning subscriptions in poll(), cache the result during metadata updates
     private boolean cachedSubscriptionHasAllFetchPositions;
     private final WakeupTrigger wakeupTrigger = new WakeupTrigger();
+    private final AtomicBoolean fencedInstance = new AtomicBoolean(false);
```
```suggestion
    private volatile boolean isFenced = false;
```

Any reason this needs to be anything other than a `boolean`?

kirktrue reviewed (changes requested):
Thanks for the updates, @philipnee. I can appreciate the game of Twister you're having to play to get this to work.

Take a look at the forthcoming [changes to `PrototypeAsyncConsumer` in my `groupMetadata` pull request](https://github.com/kirktrue/kafka/blob/KAFKA-15281-implement-groupmetadata/clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java#L1115). With those changes, I would imagine that you could add a new `BackgroundEvent` and enqueue it in the background event queue in your callback in `commitAsync`:

```java
public void commitAsync(Map<TopicPartition, OffsetAndMetadata> offsets, OffsetCommitCallback callback) {
    CompletableFuture<Void> future = commit(offsets, false);
    future.whenComplete((r, t) -> {
        // Note: this Future is being completed by the background thread, so only enqueue an event
        // for the application thread
        backgroundEventHandler.add(new OffsetCommitBackgroundEvent(callback, offsets, t));
    });
}
```

Then you'd update the inner `BackgroundEventProcessor` to add a new `process` method like this:

```java
private void process(final OffsetCommitBackgroundEvent event) {
    Exception exception = event.exception();
    OffsetCommitCallback callback = event.callback();
    Map<TopicPartition, OffsetAndMetadata> offsets = event.offsets();

    if (exception instanceof RetriableException) {
        if (callback != null)
            callback.onComplete(offsets, new RetriableCommitFailedException(exception));
        else
            log.error("Offset commit with offsets {} failed", offsets, t);

        return;
    }

    if (exception instanceof FencedInstanceIdException)
        fencedInstance = true;

    if (callback != null)
        callback.onComplete(offsets, exception);
    else
        log.error("Offset commit with offsets {} failed", offsets, t);
}
```

The `BackgroundEventProcessor` is already invoked inside `poll()` and if the event is added to the queue and the `process` method is added similar to the above, that should do the trick.

philipnee committed:
addressed comments

philipnee committed:
Update PrototypeAsyncConsumer.java

lucasbru commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java:
```diff
@@ -147,6 +151,10 @@ public class PrototypeAsyncConsumer<K, V> implements Consumer<K, V> {
     // to keep from repeatedly scanning subscriptions in poll(), cache the result during metadata updates
     private boolean cachedSubscriptionHasAllFetchPositions;
     private final WakeupTrigger wakeupTrigger = new WakeupTrigger();
+    private volatile boolean isFenced = false;
```
I'm wondering why we need this member. It looks like the volatile boolean is written and read only from the application thread. Could it then just be a boolean? But if so, could we just throw the exception directly and avoid the indirection?

philipnee replied:
Thanks, we don't need this to be volatile.  In terms of exception throwing, I think once the instance is fenced, the we want to keep throwing if the user tries to commit or poll, because the expectation is that the user should close and reconfigure the instance.  See the original implementation

```
void invokeCompletedOffsetCommitCallbacks() {
        if (asyncCommitFenced.get()) {
            throw new FencedInstanceIdException("Get fenced exception for group.instance.id "
                + rebalanceConfig.groupInstanceId.orElse("unset_instance_id")
                + ", current member.id is " + memberId());
        }
         ...
}
```

Here asyncCommitFenced is set during the callback failure in the original implementation.
```
if (commitException instanceof FencedInstanceIdException) {
                    asyncCommitFenced.set(true);
}
```

lucasbru replied:
Make sense, thanks!

lucasbru replied:
@philipnee Can we remove the volatile if we do not need this?

lucasbru commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java:
```diff
@@ -278,7 +286,8 @@ public PrototypeAsyncConsumer(final Time time,
 
-    public PrototypeAsyncConsumer(LogContext logContext,
+    // Visible for testing
+    PrototypeAsyncConsumer(LogContext logContext,
```
You will have to fix the indentation of the other parameters.

lucasbru commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java:
```diff
@@ -381,21 +391,28 @@ public void commitAsync(OffsetCommitCallback callback) {
     }
 
     // Visible for testing
     CompletableFuture<Void> commit(Map<TopicPartition, OffsetAndMetadata> offsets, final boolean isWakeupable) {
```
nit: some parameters final, some non-final

lucasbru commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java:
```diff
@@ -381,21 +391,28 @@ public void commitAsync(OffsetCommitCallback callback) {
+
+        if (offsets.isEmpty()) {
+            return CompletableFuture.completedFuture(null);
+        }
```
Maybe not related to this concrete PR, but the original commit code has some logic around updating the last seen leader epochs here, can we omit it, or is it a follow-up task?

philipnee replied:
Thanks, I'll add a PR for investigation. I think metadata was not in the implementation at the time the function was implemented.

philipnee replied:
Here is the Jira: https://issues.apache.org/jira/browse/KAFKA-15803

lucasbru commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java:
```diff
@@ -1010,15 +1031,9 @@ private void updateLastSeenEpochIfNewer(TopicPartition topicPartition, OffsetAnd
-    }
-
     boolean updateAssignmentMetadataIfNeeded(Timer timer) {
+        maybeInvokeCallbacks();
```
Sometimes we check for the fencedInstance first, sometimes we invoke the callbacks first.

I think in a previous version of the PR and in the original consumer we invoke at the beginning of `poll`. Why did you move this into `updateAssignmentMetadataIfNeeded`?

philipnee replied:
Hi, to your first question, it was an oversight, so i fixed it. to your second question: I moved it down for a couple of reasons, one is we want the callback invocation to be included in the poll timer, and two, i wanted to consolidate the logic with the backgroundEventProcess.poll to make its behavior similar to the current consumer.  The current consumer poll `ConsumerCoordinator` class in the `updateAssignmentMetadataIfNeeded`.  

the `ConsumerCoordinator#poll` actually performs the callback invocation first, which, first checks the fencedInstaceId. See the code below:

```
public boolean poll(Timer timer, boolean waitForJoinGroup) {
        maybeUpdateSubscriptionMetadata();

        invokeCompletedOffsetCommitCallbacks();
        ...
}
```

and 
```
void invokeCompletedOffsetCommitCallbacks() {
        if (asyncCommitFenced.get()) {
            throw new FencedInstanceIdException("Get fenced exception for group.instance.id "
                + rebalanceConfig.groupInstanceId.orElse("unset_instance_id")
                + ", current member.id is " + memberId());
        }
        while (true) {
            OffsetCommitCompletion completion = completedOffsetCommits.poll();
            if (completion == null) {
                break;
            }
            completion.invoke();
        }
    }
```

philipnee replied:
The `backgroundEventProcessor.process` is very similar to `ConsumerCoordinator#poll`; however, because of the implementation difference, the two implementations will differ slightly.  The change here is to ensure the two behave as similar as possible.

lucasbru commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java:
```diff
@@ -1096,4 +1111,76 @@ private void subscribeInternal(Collection<String> topics, Optional<ConsumerRebal
+    }
+
+    // Visible for testing
+    void maybeInvokeCallbacks() {
```
nit: how about `maybeInvokeCommitCallbacks`?

philipnee replied:
great!

lucasbru commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java:
```diff
@@ -1096,4 +1111,76 @@ private void subscribeInternal(Collection<String> topics, Optional<ConsumerRebal
+
+    /**
+     * Utility class that helps the application thread to invoke user registered {@link OffsetCommitCallback}. This is
+     * achieved by having the background thread to register a {@link OffsetCommitCallbackTask} to the invoker upon the
```
```suggestion
     * achieved by having the background thread register a {@link OffsetCommitCallbackTask} to the invoker upon the
```

lucasbru commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java:
```diff
@@ -1096,4 +1111,76 @@ private void subscribeInternal(Collection<String> topics, Optional<ConsumerRebal
+        }
+
+        public void executeCallbacks() {
+            LinkedList<OffsetCommitCallbackTask> callbacks = new LinkedList<>();
```
Why do we need to create a copy of the queue here? Are we concerned about consumer invocations inside the callback? Then maybe a put a comment here.

philipnee replied:
Thanks, I checked the current implementation, we don't need to do this.

lucasbru commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java:
```diff
@@ -1096,4 +1111,76 @@ private void subscribeInternal(Collection<String> topics, Optional<ConsumerRebal
+        }
+    }
+
+    private class OffsetCommitCallbackTask {
```
Should be a static class.

philipnee replied:
Hi, it is not static for the reason of the use of isFenced.  Let me know what do you think.

lucasbru commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java:
```diff
@@ -1096,4 +1111,76 @@ private void subscribeInternal(Collection<String> topics, Optional<ConsumerRebal
+     * achieved by having the background thread to register a {@link OffsetCommitCallbackTask} to the invoker upon the
+     * future completion, and execute the callbacks when user polls/commits/closes the consumer.
+     */
+    private class OffsetCommitCallbackInvoker {
```
Should be a static class.

philipnee replied:
Similar to OffsetCommitCallbackTask

lucasbru commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java:
```diff
@@ -1096,4 +1111,76 @@ private void subscribeInternal(Collection<String> topics, Optional<ConsumerRebal
+            try {
+                callbackQueue.offer(callback);
+            } catch (Exception e) {
+                log.error("Failed to enqueue OffsetCommitCallback", e);
```
Why do we expect an exception inside offer?

lucasbru commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/PrototypeAsyncConsumer.java:
```diff
@@ -712,6 +729,10 @@ private void close(Duration timeout, boolean swallowException) {
 
+        // Invoke all callbacks after the background thread exists in case if there are unsent async
+        // commits
+        maybeInvokeCallbacks();
```
Why do we not check for fenced exceptions here?

The original consumer attempts to complete all in-flight async commit requests within `timeout` ms. Will closing the background thread take care of this?

philipnee replied:
When the consumer is closing, we don't really need to throw a fenced id exception, because when the consumer is fenced we want the consumer to close right?

To your second question: yes, the work is inflight, KAFKA-15327.

lucasbru replied:
The fenced exception would indicate here that the last commit did not complete, correct? I think you are right, throwing the exception is not strictly required for correctness here; at the same time, when investigating what was going on the client during a fencing situation, it would be valuable information that the commit failed due to fencing, even if we are shutting down. If we can tell that from non-DEBUG logs, I'd say it's fine though, I haven't checked.

lucasbru reviewed (commented):
Thanks for the PR! I left some questions / comments.

philipnee committed:
PR comment updates

philipnee commented:
Hi @lucasbru - Thanks for your feedback, the PR was updated.  Would you have time to go over the comments again? Thanks!

lucasbru commented:
Hi @philipnee. Looks good, I think all my comments got addressed. I see a comment from @kirktrue about using the background event queue instead of a separate invoker queue, did we resolve this conversation? It seems to make a lot of sense to me, if at all possible.

kirktrue commented:
@lucasbru As @philipnee said, the two mechanisms appear similar and do have overlap. However, not only is it "awkward" to mix them, but it raises correctness questions.

`poll()` processes  _all_ of the events in the queue from the background thread. However, when the user calls `commit`*, we want to execute only the commit callbacks. As a result, we would have to _filter and remove_ the relevant events from the background thread.

The questions that arose:

* What if, when processing the commit callbacks, we see error events in the queue? Do we skip them? Execute them? Or something else?
* Will removing the events from the background event queue cause unforeseen issues later when calling `poll()`?

The answer may be that it's totally fine. We kind of "punted" by doing it this way, not only because of those questions, but because it seemed a bit awkward 🤷‍♂️

lucasbru commented:
Okay. Sounds like using the background events queue wouldn't strictly have downsides, but also not have so many upsides. I agree that picking things from a queue sounds awkward, so let's stick with a separate queue then

philipnee committed:
Volatile is not needed.

philipnee committed:
We need to catch the exception from the queue.offer

Update AsyncKafkaConsumer.java

philipnee commented:
Hi @lucasbru - Addressed your comment about the volatile.  I also modify the invoker.submit to catch some java exception.  Please review.  Terrible sorry but I needed to rebase to the current trunk so here are the two new commits: [8c133b8](https://github.com/apache/kafka/pull/14680/commits/8c133b87d6c8cd6b87d009c91d6955e2d23c01e5) and [78b3cfa](https://github.com/apache/kafka/pull/14680/commits/78b3cfa6710f9287b3ff70f50ce911052b4ff313)

lucasbru commented:
@philipnee there is a suspicious test failure for `org.apache.kafka.clients.consumer.KafkaConsumerTest`, could you have a look?

philipnee commented:
hm - why is it sending heartbeat? looking

philipnee committed:
Update ConsumerNetworkThread.java

Update ConsumerNetworkThread.java

philipnee commented:
I rebased the trunk and was unable to reproduce the failure locally.  I wonder if the tests is flaky and was caused by the delegate PR.

philipnee committed:
Disalbe invalid tests

disable heartbeat

philipnee commented:
I think the test failures was probably caused by heartbeat bubbled up to the consumer subsequently throwing an exception

philipnee commented:
The most recent build passed - let me re-trigger the tests to make sure.

philipnee committed:
Update KafkaConsumerTest.java

Update KafkaConsumerTest.java

philipnee committed:
Update KafkaConsumerTest.java

Update KafkaConsumerTest.java

philipnee commented:
hi @lucasbru - I think the KafkaConsumerTest failures were caused by some test we forgot to disabled.  The last 2 runs seem to be ok (well, still flaky but not failing on that specific one).  I retriggerred it again.  But I think we can merge it.

lucasbru commented:
Scala compilation in 2.12 is broken on master, which is fixed in https://github.com/apache/kafka/pull/14786, otherwise this is looking good

philipnee commented:
Thanks @lucasbru

lucasbru committed:
Merge remote-tracking branch 'upstream/trunk' into ctr-KAFKA-15562-ensure-commit-api-throws-the-right-exception

lucasbru commented:
Merged trunk. Test failures are unrelated
